File: golang/76. 最小覆盖子串.go
func minWindow(s string, t string) string {
    need := make(map[byte]int)
    window := make(map[byte]int)
    left, right := 0, 0
    valid := 0
    start, length := 0, math.MaxInt32
    for i := range t {
        need[t[i]]++
    } 
    for right < len(s) {
        c := byte(s[right])
        right++
        if _, ok := need[c]; ok {
            window[c]++
            if window[c] == need[c] {
                valid++
            }
        }
        for left < right && valid == len(need) {
            if length > right - left {
                start = left
                length = right - left
            }
            d := s[left]
            left++
            if _, ok := need[d]; ok {
                if need[d] == window[d] {
                    valid--
                }
                window[d]--
            }
        }
    }
    if length == math.MaxInt32 {
        return ""
    } else {
        return s[start : start+length]
    }
}

File: 刷题/CodeTop/76. 最小覆盖子串.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> need;
        unordered_map<char,int> window;
        int valid = 0;
        for(char c:t) {
            need[c]++;
        }
        int left = 0, right = 0;
        int start = 0, len = INT_MAX;
        while(right < s.length()) {
            char c = s[right];
            right++;
            window[c]++;
            if(need.count(c) && window[c] == need[c]) {
                valid++;
            }
            while(left < right && valid == need.size()) {
                if(right - left < len) {
                    start = left;
                    len = right - left;
                }
                char d = s[left];
                left++;
                if(need.count(d) && window[d] == need[d]) {
                    valid--;
                }
                window[d]--;
            }
        }
        return len == INT_MAX?"":s.substr(start, len);
    }
};